docs/architecture/tls13-experimental.md

```diff
@@ -66,3 +66,245 @@ together with their level of testing:
+  client_certificate_type        no    no
+  server_certificate_type        no    no
+  padding                        no    no
+  key_share                      YES   YES
+  pre_shared_key                 no    YES
+  psk_key_exchange_modes         no    YES
+  early_data                     no    YES
+  cookie                         no    YES
+  supported_versions             YES   YES
+  certificate_authorities        no    no
+  post_handshake_auth            no    no
+  signature_algorithms_cert      no    no
+
+- Supported groups: depends on the library configuration.
+  Potentially all ECDHE groups:
+  secp256r1, secp384r1, secp521r1(0x0019), x25519, x448.
+
+- Supported signature algorithms: depends on the library configuration.
+  Potentially:
+  ecdsa_secp256r1_sha256, ecdsa_secp384r1_sha384, ecdsa_secp521r1_sha512,
+  rsa_pss_rsae_sha256.
+
+- Supported versions: only TLS 1.3
+
+- Support of Mbed TLS SSL/TLS related (not DTLS) features:
+
+  The TLS 1.3 MVP is compatible with all TLS 1.2 configuration options in the
+  sense that when enabling the TLS 1.3 MVP in the library there is no need to
+  modify the configuration for TLS 1.2. Mbed TLS SSL/TLS related features are
+  not supported or not applicable to the TLS 1.3 MVP:
+
+                                         Supported  Comment
+  MBEDTLS_SSL_ALL_ALERT_MESSAGES           no
+  MBEDTLS_SSL_ASYNC_PRIVATE                no
+  MBEDTLS_SSL_CONTEXT_SERIALIZATION        no
+  MBEDTLS_SSL_DEBUG_ALL                    no
+  MBEDTLS_SSL_ENCRYPT_THEN_MAC             n/a
+  MBEDTLS_SSL_EXTENDED_MASTER_SECRET       n/a
+  MBEDTLS_SSL_KEEP_PEER_CERTIFICATE        no
+  MBEDTLS_SSL_RENEGOTIATION                n/a        Not TLS 1.2 dependent
+  MBEDTLS_SSL_MAX_FRAGMENT_LENGTH          no
+  MBEDTLS_SSL_ALPN                         no
+
+  MBEDTLS_SSL_SESSION_TICKETS              no
+  MBEDTLS_SSL_EXPORT_KEYS                  no         Incomplete support
+  MBEDTLS_SSL_SERVER_NAME_INDICATION       no
+  MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH       no
+
+  MBEDTLS_ECP_RESTARTABLE                  no
+  MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED     no
+
+  MBEDTLS_KEY_EXCHANGE_PSK_ENABLED         n/a        Make sense in TLS 1.3
+  MBEDTLS_KEY_EXCHANGE_DHE_PSK_ENABLED     n/a        context but their current
+  MBEDTLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED   n/a        definition is TLS 1.2 only.
+  MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED     n/a
+  MBEDTLS_KEY_EXCHANGE_RSA_ENABLED         n/a
+  MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED     n/a
+  MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED   n/a
+  MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED n/a
+  MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED  n/a
+  MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED    n/a
+  MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED     n/a
+
+  MBEDTLS_USE_PSA_CRYPTO                   no
+
+Not in the plan yet but probably necessary for a viable client:
+- server_name extension
+- support for HelloRetryRequest
+- fallback to TLS 1.2
```

Comment:
I think we should be more precise with the words we use here, as there are two distinct things.

1. The client sends a 1.3 ClientHello, but the server only supports 1.2, so it replies with a 1.2 ServerHello, the client notices and a 1.2 handshake is completed successfully (unless there are other issues). That's usually called version negotiation and it's a feature of the protocol. That's the thing that's currently exclude from the MVP.
2. The client sends a 1.3 ClientHello, but the server only supports 1.2, so it replies with a 1.2 ServerHello, the client notices and breaks the handshake; the application can then opt to start a new handshake using 1.2 directly, which will complete successfully (unless there are other issues). That's what's usually called fallback, and it's not part of the protocol. People have been doing that for a long time work around servers (or middle boxes) that did not implement version negotiation properly, known as version-intolerant servers (tough fortunately, this became less of an issue lately, and I think this work-around was removed from most modern browsers in the past couple of years).

The second scenario (out-of-protocol fallback by the application) is possible with the current definition of MVP. We might want to document that, if only to include a warning that, being out-of-protocol, it's not covered by the usual security guarantees of the protocol: an active attacker could easily force a downgrade to 1.2 undetected (while in-protocol negotiation is secure against active attackers). It's up to the user to decide whether that's an acceptable risk. 

It's also possible with the current definition of MVP to use TLS 1.2 with most servers and 1.3 only with servers on a "known-good" list. In closed or tightly-controlled environments (common in the embedded world), again you might just know that the servers you're talking to are all supposed to support 1.3 so neither negotiation nor fallback is needed.

So, overall while I agree not supporting version negotiation is a serious limitation, it's still a limitation that can be worked around (if the security implications are acceptable) and that's even a non-problem in some circumstances, so I don't think it would make the library unusable.

OTOH, supporting proper version negotiation does not look entirely trivial. (Currently we have separate state machines for 1.3 and 1.2 as well as separate writing and parsing functions, we'd need to merge the initial steps of those machines, and handle parsing of the ServerHello in a special way.) Also, it's something that's not part of the prototype IIRC: last time I heard, the prototype could not even build with both 1.3 and 1.2 enabled. So, it would probably be a significant effort to add support for that.

I agree that's one of the most desirable features to have beyond the current scope of MVP, but I'm not sure we should add it to the MVP EPIC: it has significant complexity/risk, and it sure adds a lot of value, but there would still be non-zero value an MVP without it IMO.

Reply:
As discussed otherwise, not in the scope of the MVP.